### ISSUE TITLE
torch.nn.functional.linear works wrong in distributed scenario.

### ISSUE THREAD

**Issue by 480284856** [pytorch/pytorch#121962]
### 🐛 Describe the bug

The output result of ```torch.nn.functional.linear``` is not equivalent with the result of that calculating the GEMM in a distributed way(like [RowParallelLinear](https://github.com/facebookresearch/fairscale/blob/9a173bf276892d8be3f12bd9e70007df5bdc4cf6/fairscale/nn/model_parallel/layers.py#L299C7-L299C24)). The formular can describe what I mean:

```math
X = [X_1, X_2] \\
```
```math
A=\begin{bmatrix} A_1 \\ A_2 \end{bmatrix}
```
```math
XA \neq X_1A_1+X_2A_2
```

### reproduce script
```python
import torch
import torch.nn.functional as F

torch.set_printoptions(20)
inputs=torch.randn(2,6, dtype=torch.float16)
weights=torch.randn(2,6, dtype=torch.float16)

outputs=F.linear(
    inputs,
    inputs
)

input1, input2=inputs.split(3,dim=1)
weight1, weight2=inputs.split(3, dim=1)

dist_output=F.linear(input1, weight1)+F.linear(input2, weight2)
print(
    dist_output==outputs,'\n',
    outputs,'\n',
    dist_output, '\n'
)
```
in my case the outputs can be(yes, the positions of unequivalent pairs are random ):
```
tensor([[False, False],
        [False,  True]]) 
 tensor([[**16.29**687500000000000000,  **0.6806**6406250000000000],
        [ **0.6806**6406250000000000,  5.73437500000000000000]],
       dtype=torch.float16)
 tensor([[**16.28**125000000000000000,  **0.6801**7578125000000000],
        [ **0.6801**7578125000000000,  5.73437500000000000000]],
       dtype=torch.float16)
```
or
```
tensor([[ True,  True],
        [ True, False]]) 
 tensor([[1.10253906250000000000, 1.16699218750000000000],
        [1.16699218750000000000, **5.855**46875000000000000]], dtype=torch.float16)
 tensor([[1.10253906250000000000, 1.16699218750000000000],
        [1.16699218750000000000, **5.859**37500000000000000]], dtype=torch.float16)
```

you can also change the ```dtype``` to ```torch.float64``` or ```torch.float```, and the results are same.

### Versions

python=3.9.18
torch==2.2.0

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by albanD** [comment]
That sounds like expected loss of precision for fp16 matmuls no?

There is https://pytorch.org/docs/stable/notes/numerical_accuracy.html that has some related details as well.

**Comment by mikaylagawarecki** [comment]
Closing since this behavior is expected, @480284856 feel free to comment if you believe otherwise

**Comment by nobody81127** [comment]
I applied a floating-point error analysis algorithm to calculate the error bounds for the result of outputs, and I found that the result of dist_output falls within this range. Therefore, I believe this is indeed a floating-point error, likely caused by differences in the order of operations. @480284856
